[multistage] Initial (phase 1) Query runtime for window functions with ORDER BY within the OVER() clause

somandal opened the pull request:
This PR introduces the multi-stage runtime changes to support Phase 1 of Window Functions, specifically targeting queries of the type `OVER(ORDER BY)` and `OVER(PARTITION BY key1 ORDER BY key2)`. Runtime support has been added for the following classes of window queries:

- `OVER(ORDER BY)` - single and multiple `OVER(ORDER BY)` in the same query
- `OVER(PARTITION BY key1 ORDER BY key2)` - single and multiple `OVER(PARTITION BY key1 ORDER BY key2)` using the same `PARTITION BY` key1 and `ORDER BY` key2 across all `OVER()s` in the same query

The window functions supported as part of Phase 1 are: SUM, AVG, MIN, MAX, COUNT, BOOL_OR, and BOOL_AND.

- Design document: https://docs.google.com/document/d/13CmFm4djI09JKF_Xty5acoXxJoxC9CLXAsmikgzgtIs/edit?usp=sharing
- Window functions Issue: https://github.com/apache/pinot/issues/7213
- Window Function Planner PR: https://github.com/apache/pinot/pull/10228
- Window Functions Runtime PR for empty `OVER()` and `OVER(PARTITION BY)`: https://github.com/apache/pinot/pull/10286
- Adding ordering as a part of the SortExchange: https://github.com/apache/pinot/pull/10408

This PR does not include support for:

- Custom frames
- Other window functions related to rank and values
- Multiple window groups (basically multiple OVER clauses with different PARTITION BY, ORDER BY and/or FRAME specifications, if these specifications are the same they get grouped into a single window group)

The above will be part of future changes for window function support

**Note:** One issue was found while working on this PR related to Apache Calcite parsing. Details can be found in [this section of the design document](https://docs.google.com/document/d/13CmFm4djI09JKF_Xty5acoXxJoxC9CLXAsmikgzgtIs/edit#heading=h.s6az8ttq3x5f). Look for heading **"Calcite Parser Exception on ORDER BY with > 1 column using default frame"**

cc @siddharthteotia @walterddr @vvivekiyer @ankitsultana

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -239,16 +238,18 @@ private boolean consumeInputBlocks() {
         return true;
       }
 
+      Key emptyOrderKey = AggregationUtils.extractEmptyKey();
```
i dont think we need to extract this on every block

somandal replied:
good catch, fixed

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -239,16 +238,18 @@ private boolean consumeInputBlocks() {
-        //       handling for all query types
+        // TODO: Revisit null direction handling for all query types
         Key key = AggregationUtils.extractRowKey(row, _groupSet);
+        Key orderKey = (CollectionUtils.isEmpty(_orderSetInfo.getOrderSet()) || _orderSetInfo.isPartitionByOnly())
```
same here. please dont do checks inside row-loop

somandal replied:
good catch, fixed

walterddr reviewed (commented):
looks good mostly.
this PR assumes the mailbox receive will generate ordered block appropriately, correct? if so could we please state so in the javadoc of window aggregate operator?

somandal commented:
> this PR assumes the mailbox receive will generate ordered block appropriately, correct? if so could we please state so in the javadoc of window aggregate operator?

Thanks for the review! Yes, this PR assumes that mailbox receive will generate ordered blocks. Added a note in the javadocs about this. I've also addressed all other comments. PTAL when you get a chance.

walterddr reviewed (approved):
lgtm
